kpi/paginators.py

```diff
@@ -43,6 +44,25 @@ def get_paginated_response(self, data, metadata):
 
         return Response(response)
 
+    @staticmethod
+    @cache_for_request
+    def get_all_asset_ids_from_queryset(queryset):
+        # Micro optimization, cast `asset_ids` as a list to force the query
+        # to be processed right now. Otherwise, because queryset is lazy query
+        # it creates (left) joins on tables when queryset is interpreted which is
+        # slower than running this extra query.
+        #
+        # AssetPagination.get_count()` and `AssetViewSet.get_serializer_context()`
+        # use the same query base. Retrieve them here and cache the results
+        asset_ids = list(queryset.values_list('id', flat=True).distinct().order_by())
```

Comment:
If you look at this line https://github.com/kobotoolbox/kpi/pull/2811/files#diff-ed6c9e5ad56864c282e4bfaca5100279R489, `asset_ids` must be coerced as a list to avoid `asset_id__in=asset_ids` to be converted as a subquery and (as I said in the comment) and to create left join on tables.